File: cli/src/client/pda.rs
use anchor_client::solana_sdk::pubkey::Pubkey;

pub fn find_farm_address<T: AsRef<[u8]>>(
    farm_authority: T,
    reward_mint: T,
) -> Pubkey {
    Pubkey::find_program_address(
        &[b"farm", farm_authority.as_ref(), reward_mint.as_ref()],
        &dskullys_staking::id(),
    )
    .0
}

pub fn find_farm_manager_address<T: AsRef<[u8]>>(
    farm: T,
    authority: T,
) -> Pubkey {
    Pubkey::find_program_address(
        &[b"farm_manager", farm.as_ref(), authority.as_ref()],
        &dskullys_staking::id(),
    )
    .0
}

pub fn find_whitelist_proof_address<T: AsRef<[u8]>>(
    farm: T,
    address: T,
) -> Pubkey {
    Pubkey::find_program_address(
        &[b"collection_data", farm.as_ref(), address.as_ref()],
        &dskullys_staking::id(),
    )
    .0
}

#[allow(dead_code)]
pub fn find_stake_receipt_address<T: AsRef<[u8]>>(
    farmer: T,
    mint: T,
) -> Pubkey {
    Pubkey::find_program_address(
        &[b"stake_receipt", farmer.as_ref(), mint.as_ref()],
        &dskullys_staking::id(),
    )
    .0
}

File: app/lib/pda.ts
import { utils } from "@project-serum/anchor";
import { PublicKey } from "@solana/web3.js";

import { PROGRAM_ID } from "./gen/programId";

export const findFarmAddress = ({
  authority,
  rewardMint,
}: {
  authority: PublicKey;
  rewardMint: PublicKey;
}): PublicKey =>
  utils.publicKey.findProgramAddressSync(
    [Buffer.from("farm"), authority.toBuffer(), rewardMint.toBuffer()],
    PROGRAM_ID
  )[0];

export const findFarmerAddress = ({
  farm,
  owner,
}: {
  farm: PublicKey;
  owner: PublicKey;
}): PublicKey =>
  utils.publicKey.findProgramAddressSync(
    [Buffer.from("farmer"), farm.toBuffer(), owner.toBuffer()],
    PROGRAM_ID
  )[0];

export const findFarmManagerAddress = ({
  farm,
  authority,
}: {
  farm: PublicKey;
  authority: PublicKey;
}): PublicKey =>
  utils.publicKey.findProgramAddressSync(
    [Buffer.from("farm_manager"), farm.toBuffer(), authority.toBuffer()],
    PROGRAM_ID
  )[0];

export const findWhitelistProofAddress = ({
  farm,
  creatorOrMint,
}: {
  farm: PublicKey;
  creatorOrMint: PublicKey;
}): PublicKey =>
  utils.publicKey.findProgramAddressSync(
    [Buffer.from("collection_data"), farm.toBuffer(), creatorOrMint.toBuffer()],
    PROGRAM_ID
  )[0];

export const findStakeReceiptAddress = ({
  farmer,
  mint,
}: {
  farmer: PublicKey;
  mint: PublicKey;
}): PublicKey =>
  utils.publicKey.findProgramAddressSync(
    [Buffer.from("stake_receipt"), farmer.toBuffer(), mint.toBuffer()],
    PROGRAM_ID
  )[0];
